pkg/apis/config/artifact_bucket.go

```diff
@@ -0,0 +1,112 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// bucketLocationKey is the name of the configmap entry that specifies
+	// loction of the bucket.
+	BucketLocationKey = "location"
+
+	// BucketServiceAccountSecretNameKey is the name of the configmap entry that specifies
+	// the name of the secret that will provide the servie account with bucket access.
+	// This secret must  have a key called serviceaccount that will have a value with
+	// the service account with access to the bucket
+	BucketServiceAccountSecretNameKey = "bucket.service.account.secret.name"
+
+	// BucketServiceAccountSecretKeyKey is the name of the configmap entry that specifies
+	// the secret key that will have a value with the service account json with access
+	// to the bucket
+	BucketServiceAccountSecretKeyKey = "bucket.service.account.secret.key"
+
+	// DefaultBucketServiceAccountSecretKey defaults to a gcs bucket
+	DefaultBucketServiceFieldName = "GOOGLE_APPLICATION_CREDENTIALS"
+
+	// bucketServiceAccountFieldName is the name of the configmap entry that specifies
+	// the field name that should be used for the service account.
+	// Valid values: GOOGLE_APPLICATION_CREDENTIALS, BOTO_CONFIG.
+	BucketServiceAccountFieldNameKey = "bucket.service.account.field.name"
+)
+
+// ArtifactPVC holds the configurations for the artifacts PVC
+// +k8s:deepcopy-gen=true
+type ArtifactBucket struct {
+	Location                 string
+	ServiceAccountSecretName string
+	ServiceAccountSecretKey  string
+	ServiceAccountFieldName  string
+}
+
+// GetArtifactBucketConfigName returns the name of the configmap containing all
+// customizations for the storage bucket.
+func GetArtifactBucketConfigName() string {
+	if e := os.Getenv("CONFIG_ARTIFACT_BUCKET_NAME"); e != "" {
+		return e
+	}
+	return "config-artifact-bucket"
+}
+
+// Equals returns true if two Configs are identical
+func (cfg *ArtifactBucket) Equals(other *ArtifactBucket) bool {
+	if cfg == nil && other == nil {
+		return true
+	}
+
+	if cfg == nil || other == nil {
+		return false
+	}
+
+	return other.Location == cfg.Location &&
+		other.ServiceAccountSecretName == cfg.ServiceAccountSecretName &&
+		other.ServiceAccountSecretKey == cfg.ServiceAccountSecretKey &&
+		other.ServiceAccountFieldName == cfg.ServiceAccountFieldName
+}
+
+// NewArtifactBucketFromMap returns a Config given a map corresponding to a ConfigMap
+func NewArtifactBucketFromMap(cfgMap map[string]string) (*ArtifactBucket, error) {
+	tc := ArtifactBucket{
+		ServiceAccountFieldName: DefaultBucketServiceFieldName,
+	}
+
+	if location, ok := cfgMap[BucketLocationKey]; ok {
```

Comment:
You might be able to save some of these checks with the helpers in knative/pkg/configmap:

```go
cm.AsString(BucketLocationKey, &tc.Location)
```

Reply:
I will check those out, thanks. I could update all config modules together in a separate patch.